src/librustc_typeck/check/_match.rs

```diff
@@ -378,12 +378,61 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
+                                        hir::Node::ImplItem(hir::ImplItem {
+                                            node: hir::ImplItemKind::Method(..), ..
+                                        }) => { // this pat is likely an argument
+                                            if let Ok(snippet) = tcx.sess.source_map()
+                                                .span_to_snippet(inner.span)
+                                            { // FIXME: turn into structured suggestion, will need
+                                              // a span that also includes the the arg's type.
+                                                err.help(&format!(
+                                                    "did you mean `{}: &{}`?",
+                                                    snippet,
+                                                    expected,
+                                                ));
+                                            }
+                                        }
+                                        hir::Node::Expr(hir::Expr {
+                                            node: hir::ExprKind::Match(..), ..
+                                        }) => { // rely on match ergonomics
+                                            if let Ok(snippet) = tcx.sess.source_map()
+                                                .span_to_snippet(inner.span)
+                                            {
+                                                err.span_suggestion(
+                                                    pat.span,
+                                                    "you can rely on match ergonomics and remove \
+                                                     the explicit borrow",
+                                                    snippet,
+                                                    Applicability::MaybeIncorrect,
+                                                );
+                                            }
+                                        }
+                                        hir::Node::Pat(_) => {  // nested `&&pat`
+                                            if let Ok(snippet) = tcx.sess.source_map()
+                                                .span_to_snippet(inner.span)
+                                            {
+                                                err.span_suggestion(
+                                                    pat.span,
+                                                    "you can probaly remove the explicit borrow",
+                                                    snippet,
+                                                    Applicability::MaybeIncorrect,
+                                                );
+                                            }
+                                        }
+                                        _ => {} // don't provide suggestions in other cases #55175
```

Comment:
I'd also put the comment here on the arm since that looks more readable when scrolling down.